src/cargo/ops/cargo_compile.rs

```diff
@@ -426,6 +425,56 @@ pub fn compile_ws<'a>(
             extra_compiler_args,
         )?;
         let mut cx = Context::new(config, &bcx)?;
+
+        // Filter to just the units whose target is activated.
+        let units = units
+            .into_iter()
+            .filter(|u| {
+                let kind = if u.target.for_host() {
+                    Kind::Host
+                } else {
+                    Kind::Target
+                };
+
+                cx.bcx.target_platform_activated(&u.target, kind)
+            })
+            .collect::<Vec<_>>();
+
+        // Ensure that there is only one [lib] activated per crate.
+        {
+            let mut pkg_libs: HashMap<(&PackageId, &str), Vec<&Unit>> = Default::default();
+
+            for unit in &units {
+                if let TargetKind::Lib(ref pkgid, _) = *unit.target.kind() {
```

Comment:
Hm, if I understand correctly, you don't have to add `pkgid` to `TargetKind`, because you can use `unit.pkg.package_id()`?

Reply:
Thanks, I didn't see that.